File: apps/server/agent/src/tools/firecrawl.ts
/**
 * Firecrawl HTTP client — keyless-first.
 * Uses https://api.firecrawl.dev/v1 without Authorization by default;
 * if FIRECRAWL_API_KEY is set, sends Bearer token. Supports self-hosted
 * via FIRECRAWL_BASE_URL.
 */

const FIRECRAWL_BASE_URL = process.env.FIRECRAWL_BASE_URL?.replace(/\/+$/, "") || "https://api.firecrawl.dev/v1";
const FIRECRAWL_API_KEY = process.env.FIRECRAWL_API_KEY;

const MAX_MARKDOWN_CHARS = 25_000;
const TRUNCATE_NOTICE = "\n\n[...content truncated...]";

function truncateMarkdown(text: string): string {
  if (text.length <= MAX_MARKDOWN_CHARS) return text;
  return text.slice(0, MAX_MARKDOWN_CHARS) + TRUNCATE_NOTICE;
}

function buildHeaders(): Record<string, string> {
  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (FIRECRAWL_API_KEY) {
    headers["Authorization"] = `Bearer ${FIRECRAWL_API_KEY}`;
  }
  return headers;
}

export interface FirecrawlSearchResult {
  title?: string;
  url?: string;
  markdown?: string;
  description?: string;
}

export interface FirecrawlSearchResponse {
  success: boolean;
  data?: FirecrawlSearchResult[];
  error?: string;
}

export async function firecrawlSearch(
  query: string,
  limit = 5,
  signal?: AbortSignal,
): Promise<FirecrawlSearchResponse> {
  const response = await fetch(`${FIRECRAWL_BASE_URL}/search`, {
    method: "POST",
    headers: buildHeaders(),
    body: JSON.stringify({
      query,
      limit,
      scrapeOptions: { formats: ["markdown"], onlyMainContent: true },
    }),
    signal,
  });

  if (!response.ok) {
    const body = await response.text().catch(() => "");
    throw new Error(`Firecrawl search error: HTTP ${response.status} ${body.slice(0, 500)}`);
  }

  const json = (await response.json()) as FirecrawlSearchResponse;
  // Truncate markdown per result
  if (json.data) {
    for (const r of json.data) {
      if (r.markdown) r.markdown = truncateMarkdown(r.markdown);
    }
  }
  return json;
}

export interface FirecrawlScrapeResponse {
  success: boolean;
  data?: {
    markdown?: string;
    metadata?: {
      title?: string;
      sourceURL?: string;
      statusCode?: number;
    };
  };
  error?: string;
}

export async function firecrawlScrape(
  url: string,
  signal?: AbortSignal,
): Promise<FirecrawlScrapeResponse> {
  const response = await fetch(`${FIRECRAWL_BASE_URL}/scrape`, {
    method: "POST",
    headers: buildHeaders(),
    body: JSON.stringify({
      url,
      formats: ["markdown"],
      onlyMainContent: true,
    }),
    signal,
  });

  if (!response.ok) {
    const body = await response.text().catch(() => "");
    throw new Error(`Firecrawl scrape error: HTTP ${response.status} ${body.slice(0, 500)}`);
  }

  const json = (await response.json()) as FirecrawlScrapeResponse;
  if (json.data?.markdown) {
    json.data.markdown = truncateMarkdown(json.data.markdown);
  }
  return json;
}

export function isRetryableFirecrawlError(err: unknown): boolean {
  const msg = err instanceof Error ? err.message : String(err);
  // 429 rate limit, network failures, 5xx
  if (msg.includes("429") || msg.includes("Rate limit")) return true;
  if (msg.includes("HTTP 5") || msg.includes("HTTP 429")) return true;
  if (msg.includes("fetch failed") || msg.includes("network")) return true;
  return false;
}

File: apps/server-go/internal/agent/tools/firecrawl.go
// Firecrawl HTTP client — keyless-first. Port of
// apps/server/agent/src/tools/firecrawl.ts. Uses
// https://api.firecrawl.dev/v1 without Authorization by default; sends a
// Bearer token when FIRECRAWL_API_KEY is set. Supports self-hosted via
// FIRECRAWL_BASE_URL.
package tools

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
)

const (
	defaultFirecrawlBaseURL = "https://api.firecrawl.dev/v1"
	firecrawlMaxMarkdown    = 25_000
	firecrawlTruncateNotice = "\n\n[...content truncated...]"
)

func truncateMarkdown(text string) string {
	if len(text) <= firecrawlMaxMarkdown {
		return text
	}
	return text[:firecrawlMaxMarkdown] + firecrawlTruncateNotice
}

func firecrawlBaseURL() string {
	if base := strings.TrimRight(os.Getenv("FIRECRAWL_BASE_URL"), "/"); base != "" {
		return base
	}
	return defaultFirecrawlBaseURL
}

func firecrawlHeaders() map[string]string {
	headers := map[string]string{"Content-Type": "application/json"}
	if key := os.Getenv("FIRECRAWL_API_KEY"); key != "" {
		headers["Authorization"] = "Bearer " + key
	}
	return headers
}

func firecrawlPost(ctx context.Context, path string, body map[string]any) (map[string]any, error) {
	raw, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, firecrawlBaseURL()+path, bytes.NewReader(raw))
	if err != nil {
		return nil, err
	}
	for k, v := range firecrawlHeaders() {
		req.Header.Set(k, v)
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if resp.StatusCode >= 400 {
		snippet := respBody
		if len(snippet) > 500 {
			snippet = snippet[:500]
		}
		return nil, fmt.Errorf("firecrawl %s error: HTTP %d %s", path, resp.StatusCode, snippet)
	}
	var out map[string]any
	if err := json.Unmarshal(respBody, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// firecrawlScrape fetches url as clean markdown via Firecrawl's /scrape
// endpoint. Returns ok=false (not an error) when Firecrawl doesn't have
// markdown for the page, so the caller can fall back to a direct fetch.
func firecrawlScrape(ctx context.Context, url string) (markdown, title, sourceURL string, statusCode int, ok bool, err error) {
	json, err := firecrawlPost(ctx, "/scrape", map[string]any{
		"url": url, "formats": []string{"markdown"}, "onlyMainContent": true,
	})
	if err != nil {
		return "", "", "", 0, false, err
	}
	success, _ := json["success"].(bool)
	data, _ := json["data"].(map[string]any)
	if !success || data == nil {
		return "", "", "", 0, false, nil
	}
	md, _ := data["markdown"].(string)
	if md == "" {
		return "", "", "", 0, false, nil
	}
	meta, _ := data["metadata"].(map[string]any)
	if t, ok := meta["title"].(string); ok {
		title = t
	}
	if s, ok := meta["sourceURL"].(string); ok {
		sourceURL = s
	}
	statusCode = 200
	if sc, ok := meta["statusCode"].(float64); ok {
		statusCode = int(sc)
	}
	return truncateMarkdown(md), title, sourceURL, statusCode, true, nil
}
